model/licensing.py

```diff
@@ -27,16 +23,85 @@
 from .hasfulltablecache import HasFullTableCache
 from .patron import Hold, Loan, Patron
 
+if TYPE_CHECKING:
+    # Only import for type checking, since it creates an import cycle
+    from ..analytics import Analytics
+
 
 class PolicyException(Exception):
     pass
 
 
-class License(Base):
+class LicenseStatus(PythonEnum):
+    preorder = "preorder"
+    available = "available"
+    unavailable = "unavailable"
+
+    @classmethod
+    def get(cls, value: str):
+        return cls.__members__.get(value.lower(), cls.unavailable)
+
+
+class LicenseFunctions:
+    identifier: str
+    checkout_url: str
+    status_url: str
+    status: LicenseStatus
+    expires: Optional[datetime.datetime]
+    checkouts_left: Optional[int]
+    checkouts_available: int
+    terms_concurrency: Optional[int]
+
+    @property
+    def is_perpetual(self) -> bool:
+        return (self.expires is None) and (self.checkouts_left is None)
+
+    @property
+    def is_time_limited(self) -> bool:
+        return self.expires is not None
+
+    @property
+    def is_loan_limited(self) -> bool:
+        return self.checkouts_left is not None
+
+    @property
+    def is_expired(self) -> bool:
+        now = utc_now()
+        return (
+            (self.expires and self.expires <= now)
+            or (self.checkouts_left is not None and self.checkouts_left <= 0)
+            or (self.status != LicenseStatus.available)
+        )
+
+    @property
+    def owned(self) -> int:
+        if self.is_expired:
+            return 0
+        elif self.is_loan_limited:
+            return self.checkouts_left
+        else:
+            return self.terms_concurrency
+
+    @property
+    def available(self) -> int:
+        if self.is_expired:
+            return 0
+        else:
+            return self.checkouts_available
+
+
+class License(Base, LicenseFunctions):
```

Comment:
Moved some of the helper functions on this class out to their own mixin class called `LicenseFunctions` so they can be applied to both `License` and `LicenseData`